src/Microsoft.CodeQuality.Analyzers.Exp/Core/Design/ValidateArgumentsOfPublicMethods.cs

```diff
@@ -0,0 +1,90 @@
+                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                compilationContext.RegisterOperationBlockAction(operationBlockContext =>
+                {
+                    // Analyze externally visible methods with reference type parameters.
+                    if (!(operationBlockContext.OwningSymbol is IMethodSymbol containingMethod) ||
+                        !containingMethod.IsExternallyVisible() ||
+                        !containingMethod.Parameters.Any(p => p.Type.IsReferenceType))
+                    {
+                        return;
+                    }
+
+                    // Bail out early if we have no parameter references in the method body. 
+                    if (!operationBlockContext.OperationBlocks.HasAnyOperationDescendant<IParameterReferenceOperation>())
+                    {
+                        return;
+                    }
+
+                    // Perform analysis of all direct/indirect parameter usages in the method to get all non-validated usages that can cause a null dereference.
+                    ImmutableDictionary<IParameterSymbol, SyntaxNode> hazardousParameterUsages = null;
+                    foreach (var operationBlock in operationBlockContext.OperationBlocks)
+                    {
+                        if (operationBlock is IBlockOperation topmostBlock)
+                        {
+                            hazardousParameterUsages = ParameterValidationAnalysis.GetOrComputeHazardousParameterUsages(topmostBlock, operationBlockContext.Compilation, containingMethod);
+                            break;
+                        }
+                    }
+
+                    if (hazardousParameterUsages != null)
+                    {
+                        foreach (var kvp in hazardousParameterUsages)
+                        {
+                            IParameterSymbol parameter = kvp.Key;
+                            SyntaxNode node = kvp.Value;
+
+                            // In externally visible method '{0}', validate parameter '{1}' is non-null before using it. If appropriate, throw an ArgumentNullException when the argument is null or add a Code Contract precondition asserting non-null argument.
+                            var arg1 = containingMethod.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat);
```

Comment:
Please use better names

Reply:
This is the standard pattern used across quite a few analyzers in our repo. I think we take either of the following approaches:

1. Add a comment with the entire diagnostic message format string above the format arguments and use names such as `arg1`, `arg2` etc. for format arguments so the reader can map the count as well as position.
2. Use better names for arguments if omitting the diagnostic message format string, so the reader can understand the code better.

I personally prefer 1. and have retained the current style.